spec/libraries/duplicator_spec.rb

```diff
@@ -0,0 +1,756 @@
+          duplicated_stuff = @duplicator.duplicate([@c1, @c21])
+
+          expect(duplicated_stuff.length).to be(2)
+          expect(duplicated_stuff[0]).to eq(@c1)
+          expect(duplicated_stuff[0]).to_not be(@c1)
+          expect(duplicated_stuff[1]).to eq(@c21)
+          expect(duplicated_stuff[1]).to_not be(@c21)
+        end
+      end
+
+      context 'when joined graphs are duplicated' do
+        before :each do
+          create_cyclic_graph
+          create_second_cyclic_graph
+          # join graphs
+          c1_children = @c1.children
+          c1_children << @c21
+          @c1.instance_variable_set(:@children, c1_children)
+
+          @duplicator = Duplicator.new
+        end
+
+        it 'duplicates all objects' do
+          @duplicator.duplicate(@c1)
+
+          duplicated_objects = @duplicator.instance_variable_get(:@duplicated_objects)
+          expect(duplicated_objects.length).to be(9)
+        end
+
+        it 'duplicates cyclically joined graphs' do
+          # join from c21 to c1
+          c21_children = @c21.children
+          c21_children << @c1
+          @c21.instance_variable_set(:@children, c21_children)
+
+          @duplicator.duplicate(@c21)
+
+          duplicated_objects = @duplicator.instance_variable_get(:@duplicated_objects)
+          expect(duplicated_objects.length).to be(9)
+        end
+      end
+    end
+  end
+
+  context 'when ActiveRecord objects' do
+    class SimpleActiveRecord < ActiveRecord::Base
+      def initialize_duplicate(_duplicator)
+      end
+    end
+
+    temporary_table(:simple_active_records) do |t|
+      t.integer :data
+    end
+
+    class ComplexActiveRecord < ActiveRecord::Base
+      has_and_belongs_to_many :children, class_name: 'ComplexActiveRecord',
+                                         foreign_key: 'parent_id',
+                                         join_table: :children_parents,
+                                         association_foreign_key: 'children_id'
+      has_and_belongs_to_many :parents, class_name: 'ComplexActiveRecord',
+                                        foreign_key: 'children_id',
+                                        join_table: :children_parents,
+                                        association_foreign_key: 'parent_id'
+
+      def initialize_duplicate(duplicator)
+        new_children = []
+
+        children.each do |child|
+          new_child = duplicator.duplicate(child)
+          new_children << new_child unless new_child.nil?
+        end
+
+        duplicator.duplicate(self).children = new_children
+      end
+    end
+
+    temporary_table(:complex_active_records) do |t|
+      t.integer :data
+    end
+
+    temporary_table(:children_parents) do |t|
+      t.integer :children_id, foreign_key: { references: :complex_active_records, primary_key: :id }
+      t.integer :parent_id, foreign_key: { references: :complex_active_records, primary_key: :id }
+    end
+
+    def create_ar_cyclic_graph
```

Comment:
Assignment Branch Condition size for create_ar_cyclic_graph is too high. [17.09/15]